src/Silverback.Integration.Kafka/Messaging/Messages/KafkaKeyHelper.cs

```diff
@@ -1,44 +1,66 @@
-                    .GetProperties()
-                    .Where(propertyInfo => propertyInfo.IsDefined(typeof(KafkaKeyMemberAttribute), true))
-                    .Select(
-                        propertyInfo => new
-                        {
-                            propertyInfo.Name,
-                            Value = propertyInfo.GetValue(message, null)?.ToString()
-                        })
-                    .ToList();
-
-            if (!keysDictionary.Any() ||
-                keysDictionary.All(keyValue => string.IsNullOrEmpty(keyValue.Value)))
-            {
-                return null;
-            }
-
-            return keysDictionary.Count == 1
-                ? keysDictionary.First().Value
-                : string.Join(",", keysDictionary.Select(p => $"{p.Name}={p.Value}"));
-        }
-    }
-}
+﻿// Copyright (c) 2020 Sergio Aquilini
+// This code is licensed under MIT license (see LICENSE file for details)
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using Silverback.Messaging.HealthChecks;
+
+namespace Silverback.Messaging.Messages
+{
+    internal static class KafkaKeyHelper
+    {
+        private static readonly Dictionary<Type, PropertyInfo[]> _typeCache = new();
+
+        public static string? GetMessageKey(object? message)
+        {
+            if (message == null)
+                return null;
+
+            if (message is PingMessage)
+                return Guid.NewGuid().ToString("N", CultureInfo.InvariantCulture);
+
+            if (!_typeCache.TryGetValue(message.GetType(), out PropertyInfo[] propertyInfos))
+            {
+                Type messageType = message.GetType();
+                propertyInfos = messageType
+                        .GetProperties()
+                        .Where(static propertyInfo => propertyInfo.IsDefined(typeof(KafkaKeyMemberAttribute), true))
+                        .ToArray();
+                _typeCache.Add(messageType, propertyInfos);
+            }
+
+            if (propertyInfos.Length == 0)
+            {
+                return null;
+            }
+
+            if (propertyInfos.Length == 1)
+            {
+                return propertyInfos[0].GetValue(message, null)?.ToString();
+            }
+
+            StringBuilder stringBuilder = new();
+            for (int i = 0; i < propertyInfos.Length; i++)
+            {
+                string name = propertyInfos[i].Name;
+                string? value = propertyInfos[i].GetValue(message, null)?.ToString();
+                if (string.IsNullOrEmpty(value))
+                {
+                    continue;
+                }
+
+                stringBuilder.Append(name);
+                stringBuilder.Append('=');
+                stringBuilder.Append(value);
+                stringBuilder.Append(',');
+            }
+
+            stringBuilder.Remove(stringBuilder.Length - 1, 1);

```

Comment:
It would be a bit faster to check if the sb is empty on line 55 and if not, add a , before add name/value.